Add TTGlyphPointPen

jenskutilek opened the pull request:
See https://github.com/googlefonts/ufo2ft/issues/469: The TTGlyphPen may change the point indices. This adds a PointPen-based TTGlyphPointPen to avoid changes in the outline.

I'm not sure what to put in the tests; how strict must path handling be? E.g. What happens if `beginPath()` is called twice in a row, or `addPoint()` is not called on a path; or should the base PointPen take care of those edge cases?

jenskutilek commented:
I added some assertions to the pen so that beginPath can not be called twice, etc. which is more strict than the normal PointPen. I also adapted the tests from TTGlyphPen.

justvanrossum commented:
Quick comments:

- consider adding it to the ttGlyphPen module (the recording pens also live in the same place)
- can both tt pens be refactored to share most code?
- could the segment pen be implemented in terms of the point pen?

anthrotype commented on Tests/pens/ttGlyphPointPen_test.py:
```diff
@@ -0,0 +1,324 @@
+from fontTools.ttLib.tables import ttProgram
+
+
+class TTGlyphPointPenTest(TestCase):
```
for new test modules we should try to use the pytest API rather then unittest. Among the nice things it provides an easy way to parametrize repetitive tests. You probably copied form the existing ttGlyphPen_test.py which still uses unittest.

jenskutilek replied:
Thanks, I changed the tests for both TTGlyphPen and TTGlyphPointPen to pytest.

anthrotype reviewed (commented):
like Just said, we should avoid duplication between the TTGlyphPen and this. Please move to the same module and refactor to reuse as much as you can.

Maybe try also reimplementing the segment-oriented pen in terms of the point pen like Just suggested earlier, e.g. by having TTGlyphPen use the SegmentToPointPen internally delegating to an instance of TTGlyphPointPen

justvanrossum commented:
Or perhaps they can have a common base class.

madig commented:
Maybe I can fuzz this in #2202 🤔 If you want to be anal about error handling (I would), check the changes I did to pens in that PR.

jenskutilek commented:
I moved the `TTGlyphPointPen` class into pens.ttGlyphPen, and also implemented both pens with a common base class, `TTGlyphBasePen`.

I also tried to subclass `TTGlyphPen` from `SegmentToPointPen` and use `TTGlyphPointPen` in there, but it turned out it gets more complex, so I kept the common subclass approach.

jenskutilek commented:
> Maybe I can fuzz this in #2202 🤔 If you want to be anal about error handling (I would), check the changes I did to pens in that PR.

Looks useful, I could add the improved error handling after your PR is merged.

anthrotype commented on Lib/fontTools/pens/ttGlyphPen.py:
```diff
@@ -47,36 +55,127 @@ def __init__(self, glyphSet, handleOverflowingTransforms=True):
         self.endPts = []
         self.types = []
         self.components = []
+        self.currentPath = None
```
self.currentPath is not used by TTGlyphBasePen nor by TTGlyphPen, but only by TTGlyphPointPen. It would make more sense to have it defined only in the latter's `__init__` method, after the `super().__init__()` has been called.

jenskutilek replied:
As TTGlyphPointPen has no `__init__` method yet, I think leaving an unused instance variable probably doesn't hurt, WDYT?

anthrotype commented on Lib/fontTools/pens/ttGlyphPen.py:
```diff
@@ -103,72 +202,79 @@ def closePath(self):
+        """
+        Returns a :py:class:`~._g_l_y_f.Glyph` object representing the glyph.
+        """
+        assert self.currentPath is None
```
`assert self.currentPath is None` could be the content of `TTGlyphPointPen._isClosed`, the point-pen version of `_isClosed` method of `TTGlyphPen`. The base class could define an "abstract" `_isClosed` that raises NotImplementedError, so the subclasses must override it with specific implementations. Then you would no longer need to have separate `_glyph` and `glyph` methods, with the latter only difference in the way to check that the countor is closed.

justvanrossum replied:
About `self.currentPath`: why is it even a list? It's effectively a three-state flag: 
1. no current path (not a valid contour)
2. there is a current path but no points have been added (not a valid contour)
3. there is a current path with at least one point (a valid contour)

Idea: call it `_currentContourStartIndex` or something, set to `None` upon reset, and set to `len(self.points)` when starting a new contour. Then you can compare with `len(self.points)` to see if there are any points.

anthrotype commented on Tests/pens/ttGlyphPointPen_test.py:
```diff
@@ -1,24 +1,25 @@
 
-class TTGlyphPointPenTest(TestCase):
-
+class TTGlyphPointPenTest:
```
you should move this test class in the same test module as ttGlyphPen_test.py, since we decided to also move the class itself to the same Lib module. The test directory has the same layout as the Lib one.

anthrotype commented on Tests/pens/ttGlyphPointPen_test.py:
```diff
@@ -112,16 +112,16 @@ def test_glyph_decomposes(self):
-        self.assertEqual(plainGlyph, compositeGlyph)
+        assert plainGlyph == compositeGlyph
 
     # def test_remove_extra_move_points(self):
```
why are these left in but commented out?

anthrotype reviewed (commented):
thanks Jens, I left some more comments.

justvanrossum commented on Lib/fontTools/pens/ttGlyphPen.py:
```diff
@@ -1,30 +1,38 @@
-    def __init__(self, glyphSet, handleOverflowingTransforms=True):
-        """Construct a new pen.
+    def __init__(
+        self, glyphSet: _TTGlyphSet, handleOverflowingTransforms: bool = True
```
This can be any kind of glyph set. I don't think it makes sense to use this (private) class as a type declaration.

anthrotype replied:
true, maybe leave it as Dict[str, Any] until we define some Drawable protocol that we can annotate this with, like https://github.com/fonttools/ufoLib2/blob/master/src/ufoLib2/typing.py

justvanrossum commented on Lib/fontTools/pens/ttGlyphPen.py:
```diff
@@ -1,30 +1,38 @@
 
         Args:
-            glyphSet (ttLib._TTGlyphSet): A glyphset object, used to resolve components.
+            glyphSet (ttLib.ttFont._TTGlyphSet): A glyphset object, used to resolve components.
```
Hm, so _TTGlyphSet was mentioned already. Doesn't make sense to me, though.

jenskutilek replied:
I changed it.

anthrotype commented on Tests/pens/ttGlyphPen_test.py:
```diff
@@ -290,6 +290,310 @@ def test_scaled_component_bounds(self):
 
 
+class TTGlyphPointPenTest:
+    def runEndToEnd(self, filename):
```
this `runEndToEnd` is basically identical to the one in `runEndToEnd` (except for the pen class) maybe have a base class that defines that and have both subclass from it, with the pen class defined as a class attribute

anthrotype commented on Tests/pens/ttGlyphPen_test.py:
```diff
@@ -290,6 +290,310 @@ def test_scaled_component_bounds(self):
+    #     pen.lineTo((0, 0))  # the duplicate point is not removed
+    #     pen.closePath()
+    #     assert len(pen.points) == 5
+    #     assert pen.points[0] == (0, 0)
```
remove commented out code or explain with a comment why it is left in

anthrotype commented on Tests/pens/ttGlyphPen_test.py:
```diff
@@ -290,6 +290,310 @@ def test_scaled_component_bounds(self):
+    def test_e2e_linesAndSimpleComponents(self):
+        self.runEndToEnd("TestTTF-Regular.ttx")
+
+    def test_e2e_curvesAndComponentTransforms(self):
```
these methods are intentical for TTGlyphPenTest and TTGlyphPointPenTest classes, use the same trick that I suggested for runEndToEnd above, and define them in the base class then subclass from them, the test will be run with both using different pen classes 

Only make sure you don't name the base class with `*Test` suffix otherwise pytest thinks it is a concrete test class (e.g. like `TTGlyphPenTest`). Call it "TTGlyphPenTestBase" maybe

anthrotype commented on Tests/pens/ttGlyphPen_test.py:
```diff
@@ -290,6 +290,310 @@ def test_scaled_component_bounds(self):
+        with pytest.raises(AssertionError):
+            pen.beginPath()
+
+    def test_closePath_ignoresAnchors(self):
```
what do you mean here by "anchors"? I only see a beginPath and an endPath (an empty contour)

justvanrossum commented on Lib/fontTools/pens/ttGlyphPen.py:
```diff
@@ -47,36 +54,136 @@ def __init__(self, glyphSet, handleOverflowingTransforms=True):
         self.endPts = []
         self.types = []
         self.components = []
+        self._currentContourStartIndex = None
```
I currently don't see this used as an index, just as a flag meaning "we're in between beginPath and endPath".

jenskutilek replied:
I moved it to the subclass.

anthrotype commented on Lib/fontTools/pens/ttGlyphPen.py:
```diff
@@ -47,36 +54,136 @@ def __init__(self, glyphSet, handleOverflowingTransforms=True):
+        glyph.flags = array("B", self.types)
+        self.init()
+
+        if components:
```
the current decomposing behaviour it's tested here

https://github.com/fonttools/fonttools/blob/31ab3aae0b3061ea106551adc54c124f8afec465/Tests/pens/ttGlyphPen_test.py#L78-L107

jenskutilek replied:
Argh, I saw that part but thought it only decomposed because of the overflowing transformation.

Should a warning be logged when a decompose happens?

jenskutilek replied:
Silently decomposing should be fine unless you absolutely want to avoid overlaps.

A theoretical TrueType hinting authoring tool — should it even allow to hint a mixed composite — will notice that the glyph has changed when compiling and hopefully throw an error.

anthrotype replied:
if you like you can add that assert and make it even clearer in the docstring that this is the expected behavior, but it's a bit too late to warn in my opinion.

jenskutilek replied:
I've added a note to the docstring.

anthrotype commented on Lib/fontTools/pens/ttGlyphPen.py:
```diff
@@ -103,72 +215,73 @@ def closePath(self):
-                continue
+class TTGlyphPointPen(TTGlyphBasePen, LogMixin, AbstractPointPen):
+    """
+    Pen used for drawing to a TrueType glyph.
```
this docstring is identical to TTGlyphBasePen's, either you make it more specific (e.g. say it's a point pen, not just a pen), or if you keep the one from the base pen, you delete this one
```suggestion
    Point pen used for drawing to a TrueType glyph.
```

jenskutilek replied:
I've added setting `_currentContourStartIndex` to the `init()` method of the point pen, which now in turn calls `super().init()`.

jenskutilek replied:
... and moved the docstrings around.

anthrotype commented on Tests/pens/ttGlyphPen_test.py:
```diff
@@ -290,6 +296,260 @@ def test_scaled_component_bounds(self):
+        assert glyph.numberOfContours == 1
+        assert glyph.endPtsOfContours == [3]
+
+    def test_glyph_errorOnCurve(self):
```
```suggestion
    def test_addPoint_errorOnCurve(self):
```

anthrotype commented on Lib/fontTools/pens/ttGlyphPen.py:
```diff
@@ -103,72 +215,73 @@ def closePath(self):
+        if self._isClosed():
+            raise PenError("Contour is already closed.")
+        if self._currentContourStartIndex == len(self.points):
+            raise PenError("Tried to end an empty contour.")
```
I wonder if it's too harsh to error out if the contour is empty. We can safely ignore it I guess, with no loss of information

jenskutilek replied:
I agree beginning and immediately ending a path is "transparent", but then I wonder in which case a script may do it ... better to err on the safe side? Likewise, we could relax the checks about `endPath` being called again before a new `beginPath`, because `beginPath` may assume ending the previous path implicitly ... I think I'd leave it like it is.

jenskutilek commented:
I think this is ready to be merged now. Any more comments @anthrotype @justvanrossum?

justvanrossum reviewed (approved):
LGTM, thanks!

anthrotype reviewed (approved):
LGTM, thank you Jens!

anthrotype commented:
I just released https://github.com/fonttools/fonttools/releases/tag/4.24.0 with the new point pen
